src/main/java/codeu/model/store/basic/IndexStore.java

```diff
@@ -66,13 +71,88 @@ public void addMapping (String word, Message message) {
   /**
     * Search function for a single word, returns a Set of messages from the Hashtable if the word is in it, otherwise return null
     */
-  public Set<Message> searchWord (String word) {
-    if (wordsHash.containsKey(word)) {
-      return wordsHash.get(word);
+  public List<Message> search (String searchTarget) {
+    if (searchTarget.contains("||")) {
+      return searchUnion(searchTarget);
+    }else if (searchTarget.contains("&&")) {
+      searchTarget = searchTarget.replaceAll(" ","");
+      return searchIntersection(searchTarget);
+    } else if (searchTarget.contains(" ")) {
+      searchTarget = searchTarget.replaceAll("\\s+", "&&");
+      return searchIntersection(searchTarget);
+    }else if (wordsHash.containsKey(searchTarget)){
+      List<Message> messageList = new ArrayList<Message>(wordsHash.get(searchTarget));
+      return sortMessageList(messageList);
     }
     return null;
   }
 
+  /**
+    * Pre: the given string contains "||"
+    * returns the union of the elements in the given searchTarget
+    */
+  public List<Message> searchUnion (String searchTarget) {
+    String[] words = searchTarget.split("\\|\\||\\s+");
+
+    if (words.length == 0) {
+      return null;
+    }
+    Set<Message> hashSetResult = new HashSet<Message>();
+    for (String word : words){
+      if (search(word) != null) {
+        Set<Message> hashSetTemp = new HashSet<Message>(search(word));
+        for (Message message : hashSetTemp){
+          hashSetResult.add(message);
+        }
+      }
+    }
+    List<Message> unionResult = new ArrayList<Message>(hashSetResult);
+    unionResult = sortMessageList(unionResult);
+    return unionResult;
+  }
+
+  /**
+    * Pre: the given string contains "&&"
+    * returns the intersection of the elements in the given searchTarget
+    */
+  public List<Message> searchIntersection (String searchTarget) {
+    if (searchTarget.indexOf("&&") == 0) {
+      searchTarget = searchTarget.substring(2);
+    }
+    String[] words = searchTarget.split("&&");
+    if (words.length == 0 || search(words[0]) == null) {
+      return null;
+    }
+    List<Message> intersectionResult = new ArrayList<Message>(search(words[0]));
+    for (int i = 1 ; i < words.length ; i++) {
```

Comment:
Any reason why this isn't also a for iterator loop like in the || case above?

Reply:
yeah, I can't initialize the result to be empty since the retainAll function compares the two list and removes items from the caller. So in order to not have a redundant call for search like mentioned, I started the loop from i = 1